src/graph.cpp

```diff
@@ -64,59 +81,141 @@ Graph::addBidirectionalEdge(const unsigned int &source, const unsigned int &dest
+                    oldDist == constants::INF ? q.insert(dest) : q.decreaseKey(dest);
+                }
+            }
         }
     }
-    UFDS ufds(vertexSet.size());
-
-    edges.sort([](const std::shared_ptr<Edge> &e1, const std::shared_ptr<Edge> &e2) {
-        return e1->getLength() < e2->getLength();
-    });
+    int exec_val = addToTour(start);
+    if (exec_val != 0) return exec_val;
+    return 0;
+}
 
-    for (const std::shared_ptr<Edge> &e: edges) {
-        if (!isSameSet(e->getOrig()->getId(), e->getDest()->getId())) {
-            e->setSelected(true);
-            linkSets(e->getOrig);
+ /**
+  * Adds a vertex to the tour structure and updates the total distance
+  * Time Complexity: O(1)
+  * @param stop - Vertex to add
+  * @return execution errors (0 if none, -1 if couldn't calculate Edge length, -2 if self-loop)
+  */
+int Graph::addToTour(std::shared_ptr<Vertex> stop) {
+    if (!tour.course.empty()){
+        if ((*(tour.course).rbegin())->getId() == stop->getId()){
+            //selfloop
+            return -2;
         }
+        double aresta = findEdge((*(tour.course).rbegin()), stop);
+        if (aresta == -1){
+            //rejeitar tour
+            return aresta;
+        }
+        tour.distance += aresta;
     }
-}*/
+    tour.course.push_back(stop);
+    return 0;
+}
 
 /**
  * @brief Iterates through the vertex set using DFS, respecting if an edge is selected or not
- * Registers the path taken in the vertex's path attribute
  * Time Complexity: O(|V|+|E|)
  * @param source - Vertex where the DFS starts
  */
-/*void Graph::dfsKruskalPath(const std::shared_ptr<Vertex> &source) {
-    source->setVisited(true);
-    for (const std::shared_ptr<Edge> &e: source->getAdj()) {
-        if (!e->getDest()->isVisited() && e->isSelected()) {
-            e->getDest()->setPath(e);
-            dfsKruskalPath(e->getDest());
+void Graph::preorderMSTTraversal(std::shared_ptr<Vertex> source){
+/*    source->setVisited(true);
+    addToTour(source);
+    for (std::shared_ptr<Edge> &e : source->getAdj()){
+        if(!(e->getDest()->isVisited()) && e->isSelected()){
+            preorderMSTTraversal(e->getDest());
         }
     }
-}*/
 
+    if (source->getId() == 0) addToTour(source);
+*/}
+
+/**
+ * Calculates an approximation of the TSP, using the triangular approximation heuristic
+ * Time Complexity: O(|E|log|E|)
+ */
+void Graph::triangularTSPTour(){
+    /*
+    -Build MST
+    -Get pre-order of the mst as vector
+    -Iterate that order getting total dist
+    */
+    tour = {0,{}};
+
+    int exec_val = prim();
+    switch (exec_val){
+        case 0:
+            printTour();
+            break;
+        case -1:
+            printf("Couldn't calculate approximation of TSP for this graph!\n");
+            break;
+        case -2:
+            printf("Course would contain self-loop!\n");
+            break;
+    }
+    //preorderMSTTraversal(findVertex(0));
```

Comment:
This is still needed, otherwise this is not the algorithm we are looking for:
![image](https://github.com/MRita443/TravellingSalesman/assets/93883163/0c01e7a9-d4ad-4f97-be1e-470a2198aefb)

We just switched the mst algorithm but we need a pre order walk nonetheless

Reply:
I thought it was already implemented midst Prim's execution, however, it wasn't a dfs nor a optimal traversal